### ISSUE TITLE
perf(cuda/quant): use float accumulators in qmv below Ampere at bits < 8

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1539]
Part of #1536. Phase 1. Depends on #1537 only if a host-side dispatch is chosen over a `__CUDA_ARCH__` guard — see the implementation note. Baseline from #1538.

The cheapest high-value item in the program: the fast path already exists in the file and is simply not selected below 8 bits.

## Context

`src/lib/mlx-cpp/patches/mlx/backend/cuda/quantized/qmm/qmv.cu:191` picks the accumulator type from the weight bit width alone:

```cpp
cuda::std::conditional_t<(bits >= 8), float, T> sums[elems_per_thread] = {};
```

At `bits < 8` the accumulator is the element type. For a bf16 checkpoint on sm_70 that is a bf16 accumulator, and Volta has no bf16 ALU — every FMA is emulated as convert-to-float, fma, convert-back. The float-accumulation specializations are already written and tested for the `bits >= 8` case (`qmv.cu:77` `dequant_fma(..., float* out)`, `qmv.cu:145` `fma_tile(..., float* out)`); nothing new needs to be authored.

Measured consequence on a V100 with `qwen3.8-27B-4bit` (see #1538): `qmv` is 14.7% of GPU time and moves ~15 GB of weights per decode step in 218 ms. That is **~70 GB/s against a 900 GB/s part — 7.7% of peak.** A 4-bit GEMV that attains 7.7% of memory bandwidth is not bandwidth-bound; it is stalled on arithmetic, which for this kernel means the emulated accumulator.

`qmv` is the decode-path kernel for essentially every quantized matmul: `quantized.cpp:238,247` route to it whenever `M * B < 8`, which is every single-stream decode step.

## Scope

Select float accumulators in `qmv` (and `gather_qmv`, same file) whenever the target is pre-Ampere, regardless of bit width. Ampere and later keep today's `bits >= 8` rule exactly.

Out of scope: changing the accumulator on sm_80+, changing the dequantization order, or touching `qmm_naive`/`qmm_sm80`.

## Implementation plan

1. **Prefer the `__CUDA_ARCH__` guard.** `qmv` is AOT-compiled, not JIT'd — the host takes a function pointer (`qmv.cu:566` `&qmv_kernel<...>`, `:611` `&gather_qmv_kernel<...>`) and launches via `add_kernel_node_raw` (`:739`, `:794`). Because nvcc compiles one device pass per entry in `MLX_CUDA_ARCHITECTURES`, a guard inside the device code gives exactly per-arch behavior with no host dispatch and no change to the host-visible template signature:
   ```cpp
   #if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 800
     using Accum = float;
   #else
     using Accum = cuda::std::conditional_t<(bits >= 8), float, T>;
   #endif
   ```
   Confirm both `dequant_fma` and `fma_tile` overload sets resolve for `(T = bfloat16_t, Accum = float)` at `bits < 8`; the `bits >= 8` path already exercises that combination, so this should be a no-op for overload resolution.
2. If the guard turns out not to be viable (e.g. a fat-binary build where one device pass would silently disagree with a host-side assumption), fall back to a `bool FloatAccum` template parameter set host-side from #1537's capability probe. Record which route was taken and why.
3. **Register pressure.** Float accumulators are 2x the accumulator register footprint of bf16. Check `-Xptxas -v` occupancy for the `<8, 16, 64>` instantiation seen in the profile and re-tune `elems_per_thread` for sm_70 only if occupancy drops enough to eat the win.
4. **Numerics.** Float accumulation is strictly more accurate than bf16 accumulation, so this improves quality — but it *does* change pre-Ampere output. Greedy decode on a Volta host will not be token-identical to the current Volta build. That is acceptable (no published Volta baseline exists) but must be stated in the changelog entry. sm_80+ output must be bit-identical; that is the regression test.

## Acceptance criteria

- [ ] `qmv` decode attains >= 25% of the 900 GB/s roofline on `qwen3.8-27B-4bit` (from 7.7%), measured as in #1538.
- [x] End-to-end decode tok/s improvement reported against the #1538 baseline of 4.2 tok/s, with the nsys kernel table before and after.
- [ ] **Zero diff on sm_80+**: greedy output byte-identical on GB10 for a fixed prompt/seed, and the emitted SASS for the sm_121 pass unchanged (`cuobjdump --dump-sass` diff, or a documented explanation if the guard perturbs it).
- [ ] `tests/cuda_qmm_determinism.rs` green on the V100: repeated greedy decode byte-identical run to run.
- [x] Quantized-matmul parity: `qmv` output vs `qmm_naive` output across bits 4/8 and group sizes 32/64/128 on representative shapes, within the tolerance those tests already use.
- [x] Changelog entry records the pre-Ampere output change explicitly.
- [ ] `cargo test --features cuda` green on both an sm_70 and an sm_121 host.

## Validation

```
MLX_CUDA_ARCHITECTURES=70 make release-cuda
for n in 8 40; do ./target/release/mlxcel generate -m ./models/qwen3.8-27B-4bit -p "Hi." -n $n; done
nsys profile -t cuda,nvtx --cuda-graph-trace=node -o qmv_after \
  ./target/release/mlxcel generate -m ./models/qwen3.8-27B-4bit -p "Explain what a GPU tensor core does." -n 24
cargo test --features cuda cuda_qmm_determinism
```

## References

- Accumulator selection: `src/lib/mlx-cpp/patches/mlx/backend/cuda/quantized/qmm/qmv.cu:191`.
- Existing float specializations to reuse: same file, `:77` and `:145`.
- Decode dispatch into `qmv`: `patches/mlx/backend/cuda/quantized/quantized.cpp:238,247,255`.
- Prior art on this file's determinism contract: #725 (multirow split, bit-identical per-row results), #910 (qmm_sm80 `cp.async` race).

**Comment by inureyes** [comment]
## Status note: premise stands, magnitude still unmeasured

The 4-bit/8-bit comparison proposed in #1538 as the empirical test for this issue **does not test it**, so this issue's expected payoff is currently supported by code reading alone, not measurement. Recording that explicitly so nobody implements against an assumed number.

Measured on a V100 with `gemma-4-26b-a4b-it-{4bit,8bit}`: 8-bit decode is 19.9% faster than 4-bit (37.63 → 31.38 ms/tok, three reps). But nsys shows that comparison is dominated by `qmm_naive`, not `qmv`:

| Arm | `qmm_naive` | `qmv` |
|---|---|---|
| 4-bit | 73.1% | 12.9% |
| 8-bit | 76.3% | 6.8% |

MoE decode routes through `GatherQMM`, where the expert count makes `M * B >= 8`, so `quantized.cpp:246-252` sends it to `qmm_naive`. Full detail in https://github.com/lablup/mlxcel/issues/1538#issuecomment-5474771994.

**What still stands unchanged** — this is a code-level fact, not an inference from that benchmark:

- `qmv.cu:191` selects the element type as accumulator at `bits < 8`; for a bf16 checkpoint that is a bf16 accumulator.
- sm_70 has no bf16 ALU, so those FMAs are emulated.
- The float-accumulation specializations already exist at `qmv.cu:77` and `:145` and are simply not selected below 8 bits.
- On the dense `qwen3.8-27B-4bit`, `qmv` attains ~70 GB/s against a 900 GB/s part (7.7%) — not a bandwidth-bound profile.

**What is missing** is the controlled measurement of how much this specific change buys. That needs a *dense* checkpoint at matched 4-bit and 8-bit, where decode is genuinely GEMV-shaped. `gemma-4-12b-it-8bit` is dense but has no 4-bit sibling locally; acquiring a matched dense pair is now a prerequisite, tracked in #1538.

No change to this issue's scope or acceptance criteria — the before/after on `qmv`'s achieved bandwidth is measured on the implementation itself and does not depend on the A/B. Only the prior expectation is unquantified.

**Comment by inureyes** [comment]
## Confirmed on a controlled dense pair: qmv is 2.14x slower at 4 bits than at 8

The premise of this issue is now measured, not just read out of the source. My previous comment said the magnitude was unquantified; it no longer is.

### The controlled pair

`mlx-community/gemma-4-12B-it-4bit` and `mlx-community/gemma-4-12b-it-8bit` on Tesla V100-PCIE-32GB. These are the same base checkpoint at two bit widths — **33 of 34 config keys are identical**, the sole difference being `"bits": 4` vs `"bits": 8` (both `group_size: 64`, `mode: affine`, `gemma4_unified`, dense, 48 layers, hidden 3840). Dense matters: unlike the MoE pair in #1538, decode here is genuinely GEMV-shaped and reaches `qmv`.

### End-to-end decode

Identical prompt, `--seed 0`, slope over `-n 40` -> `-n 120`, three repetitions:

| Rep | 4-bit ms/tok | 8-bit ms/tok |
|---|---|---|
| 1 | 124.75 | 66.63 |
| 2 | 121.50 | 62.88 |
| 3 | 119.88 | 66.88 |
| **mean** | **122.04** (8.19 tok/s) | **65.46** (15.28 tok/s) |

**8-bit decodes 1.86x faster**, while reading 12 GB of weights against 6.4 GB.

### Kernel attribution — the part that makes this conclusive

nsys `--cuda-graph-trace=node`, 120-token run, **identical launch counts in both arms**:

| Kernel | inst | 4-bit | 8-bit | Ratio |
|---|---|---|---|---|
| `qmv` (decode) | 39,151 | **12.84 s** | **5.99 s** | **8-bit 2.14x faster** |
| `qmm_naive` (prefill) | 658 | 18.23 s | 19.30 s | 4-bit 1.06x faster |

Two things fall out:

1. **The whole end-to-end difference is `qmv`.** `qmv` delta per token is (12.84 − 5.99)/120 = **57.1 ms**; the measured wall-clock slope delta is **56.6 ms**. Attribution is ~101% — nothing else is needed to explain the result.
2. **`qmm_naive` is the control that isolates the accumulator.** It accumulates in `float` via `UniversalFMA<float, Element, Element>` (`gemm_sm70.cuh:55`) *regardless of bit width*, so between these two arms it differs only in dequantization work and bytes moved — and there the 4-bit arm **wins** by 5.9%, as bandwidth would predict. The inversion appears only in the kernel where the accumulator type changes with `bits`.

That is the difference this issue exists to remove: `qmv.cu:191` selects a bf16 accumulator at `bits < 8`, and sm_70 has no bf16 ALU.

### Achieved bandwidth

| Arm | weights | qmv s/token | achieved | % of 900 GB/s |
|---|---|---|---|---|
| 4-bit | 6.4 GB | 0.107 | ~60 GB/s | **6.7%** |
| 8-bit | 12 GB | 0.050 | ~240 GB/s | **26.7%** |

The float-accumulator arm attains 4x the bandwidth fraction. Note it still only reaches 27%, so fixing the accumulator is not expected to make `qmv` bandwidth-bound — there is further headroom beyond this issue.

### Caveat on the expected magnitude

The 2.14x is an **upper bound** on what this issue alone buys. The two arms differ in dequantization work as well as accumulator type, and those cannot be separated by varying `bits`. The `qmm_naive` control suggests the dequant component is small and points the *other* way (4-bit faster by 5.9% where only dequant differs), so most of the 2.14x is plausibly the accumulator — but the honest expectation for this issue is "a large fraction of 2.14x on `qmv`", not 2.14x itself. The acceptance criterion in the issue body (>= 25% of the bandwidth roofline, from 7.7%) remains the thing to measure on the implementation.

Raw profiles and the config diff are in the audit record referenced from #1536.

**Comment by inureyes** [comment]
## Implemented in #1557: 2.01x on `qmv`, 1.87x end to end, and three criteria that stay open

Route: the `__CUDA_ARCH__ < 800` guard in device code, as the implementation plan preferred. The host-side `bool FloatAccum` fallback was not needed. `qmv` is AOT-compiled, the host only takes `&qmv_kernel<...>` as a function pointer, and the accumulator appears in neither the template signature nor the mangled name, so nvcc's per-architecture device pass gives exactly per-arch behavior with no host-visible change.

One narrowing against the scope line "regardless of bit width": the promotion is conditioned on the element type being `bfloat16_t`, not applied to every type below Ampere. sm_70 and sm_75 have a native fp16 FMA at twice the fp32 rate, so promoting f16 accumulators there would spend throughput to buy precision, the opposite of the trade this makes for bf16, and f16 policy below Ampere is #1542's subject. Every checkpoint measured here is bf16, so the narrowing costs nothing measured. The bit-width half of the rule is implemented exactly as written.

### Measured on the V100, following #1538's six rules

The before arm reproduces the committed baseline to within 1.2% on every checkpoint and to four significant figures on `qmv` kernel time, so the two records are measuring the same machine in the same state.

| Checkpoint | Before | After | Speedup |
|---|---|---|---|
| `qwen3.8-27B-4bit` | 219.96 ms/tok (4.55 tok/s) | **117.83 ms/tok (8.49 tok/s)** | **1.87x** |
| `gemma-4-12B-it-4bit` | 124.26 ms/tok (8.05 tok/s) | **71.66 ms/tok (13.95 tok/s)** | **1.73x** |
| `gemma-4-12b-it-8bit` (control) | 66.75 ms/tok | 66.73 ms/tok | 1.00x |
| `gemma-4-26b-a4b-it-4bit` (MoE) | 35.45 ms/tok | 29.56 ms/tok | 1.20x, ordinal only |

`qmv` in nsys at identical launch counts: 5.2165 to 2.5984 s over 11,431 instances (qwen, `-n 24`) and 12.8369 to 6.5654 s over 39,151 (dense, `-n 120`). Reconciliation 101.3% to 101.9% and 100.7% to 103.3%. The `qmv` delta is 52.26 ms/token against a measured slope delta of 52.60 on the dense pair, so `qmv` accounts for 99.4% of the gain. `qmm_naive`, which already accumulated in float, moves by -0.12% and +0.55%, and the 8-bit checkpoint is unmoved inside its own repeat spread. Those two controls confine the result to the accumulator.

The 4-bit against 8-bit inversion is nearly closed: the 8-bit arm's advantage falls from 1.886x to 1.074x while still reading 1.9x the bytes. Full record in `docs/benchmark_results/qmv-float-accum-v100-2026-08-31.md`, and the post-program comparison row of #1538's baseline is filled in.

### The 25%-of-roofline criterion is not met

`qmv` on `qwen3.8-27B-4bit` goes from 7.37% to **14.80%** of the 900 GB/s roofline. Reaching 25% would need another 1.69x on top of this change.

That criterion was written before the controlled measurement in the comment above existed. That measurement put the ceiling at 2.14x and said in terms that it was an upper bound, and the delivered 2.01x is 94% of it. The two figures are not consistent with each other: the 8-bit arm itself only reaches 26.7% of roofline while reading 1.9x the bytes, so no accumulator change alone could put a 4-bit arm reading half the bytes at the same fraction. The box is left unticked rather than reinterpreted. The remaining headroom belongs to the later items in #1536.

### `tests/cuda_qmm_determinism.rs` is red here, and it was red before this change

Pointed at `gemma-4-12B-it-4bit` (its default `models/llama-3.2-1b-4bit` is not present on this host or in CI, and the test skips rather than fails when the directory is missing, so every CI run of it so far has been a skip) it diverges at step 0, the prefill. That was verified against an `origin/main` build with `qmv.cu` reverted: same test, same checkpoint, same divergence at the same step. Prefill is `M = 64`, so it dispatches to `qmm_naive` and never reaches `qmv`. The divergence is between iterations inside one process rather than between the two builds, and iteration 0's prefill logits hash to the same value (5078372879123904733) on both builds, so prefill output is bit-identical across this change.

What was verified instead is the substance of the criterion: three repetitions of a greedy 32-token generation give one distinct output on the before build and one on the after build, and the two builds produce the same tokens as each other on that prompt.

**Quantized prefill on sm_70 not being bitwise reproducible is worth its own issue.** It is the same class of defect #910 fixed for `qmm_sm80` on sm_121, it is out of scope for this issue, and it is written up in the benchmark record so it is not lost.

### sm_80+ criteria: SASS verified, runtime deferred

The SASS half of the "zero diff on sm_80+" criterion **was** verified, which was not obvious to be possible here: nvcc does not need the target part to emit code for it. `qmv.cu` was compiled from both the unpatched and the patched source with the production flag set, and `cuobjdump --dump-sass` is byte-identical at sm_80 (389,998 lines) and sm_121 (437,519 lines), with `cuobjdump -res-usage` identical at both.

Byte-identical greedy output on GB10, GB10 throughput unmoved, and `cargo test --features cuda` on an sm_121 host cannot be run here, since no sm_80-or-later device exists on this machine. They are listed under `## Deferred to GB10` in #1557 and belong to the `## GB10 (sm_121) continuation` section of #1536.

### Occupancy

`cuobjdump -res-usage` at sm_70 for the `<8, 16, 64>` bf16 / uint4b instantiation: `qmv_kernel` 55 to 61 registers at `has_residue_k = false` (4 blocks/SM either way) and 59 to 66 at `has_residue_k = true` (4 blocks/SM to 3, 32 warps to 24); `qmv_multirow_kernel` 110 to 168 (2 blocks/SM to 1). Nothing spills: `STACK` and `LOCAL` are zero across all 378 `qmv`-family instantiations in both arms. The residue-k variant is the one this workload runs, so the measured 1.96x already includes its lost block per SM, and `elems_per_thread` was not re-tuned because 24 warps per SM did not come close to eating the win. The multirow kernel halves its occupancy and serves batched decode with 2 to 8 rows, which nothing measured here exercises; its row window is already autotuner-tunable (#906).

**Comment by inureyes** [comment]
## Recalibrating the roofline criterion I set, and recording it as unmet

PR #1557 reaches **14.80%** of the 900 GB/s roofline, against the `>= 25%` this issue asked for. The implementation left that box unticked rather than relaxing it, which is the right call, and the criterion was mine and was wrong.

The controlled measurement in #1538 put the ceiling for this change at 2.14x. From a 7.37% starting point that caps the arithmetic maximum near 15.8%, so 14.80% is **94% of what was achievable here**. The 25% figure came from the epic-level target in #1536 and conflated two separate costs: the 8-bit arm itself only reaches 26.7% of roofline *while reading 1.9x the bytes*, so the residual gap above ~15% is the sub-byte unpack in the inner loop, not the accumulator. Closing that is #1543's territory.

**The number to carry forward for this issue is 14.80%**, from 7.37%, a 2.01x improvement on `qmv` over identical launch counts, with 99.4% attribution of the end-to-end decode gain on the dense pair. End-to-end decode: `qwen3.8-27B-4bit` 219.96 to 117.83 ms/token (1.87x), `gemma-4-12B-it-4bit` 124.26 to 71.66 (1.73x), and the 8-bit control unchanged at 1.00x exactly as it must be.

I am not editing the acceptance criteria in the issue body after the fact. The unmet box stays visible, with this comment as the reason. The epic-level `>= 25%` target in #1536 remains open and is now correctly attributable to #1543 rather than to this issue.

One consequence for the guidance in #1538: the 4-bit against 8-bit inversion falls from 1.886x to **1.074x** on dense checkpoints. "Prefer 8-bit on Volta" was a workaround for this defect, and it is now nearly moot where this path is taken.
